File: src/main/java/net/zhuruoling/omms/central/network/session/handler/builtin/controller/ListControllersRequestHandler.java
package net.zhuruoling.omms.central.network.session.handler.builtin.controller;

import net.zhuruoling.omms.central.controller.ControllerManager;
import net.zhuruoling.omms.central.network.session.SessionContext;
import net.zhuruoling.omms.central.network.session.handler.builtin.BuiltinRequestHandler;
import net.zhuruoling.omms.central.network.session.request.Request;
import net.zhuruoling.omms.central.network.session.response.Response;
import net.zhuruoling.omms.central.permission.Permission;
import net.zhuruoling.omms.central.network.session.response.Result;
import net.zhuruoling.omms.central.util.Util;
import org.jetbrains.annotations.NotNull;

public class ListControllersRequestHandler extends BuiltinRequestHandler {
    @Override
    public Response handle(Request request, SessionContext session) {
        var controllerNames = ControllerManager.INSTANCE.getControllers().keySet();
        var json = Util.gson.toJson(controllerNames);
        return new Response().withResponseCode(Result.OK).withContentPair("names", json);
    }

    @Override
    public Permission requiresPermission() {
        return null;
    }
}

File: src/main/java/icu/takeneko/omms/central/network/session/handler/builtin/controller/ListControllersRequestHandler.kt
package icu.takeneko.omms.central.network.session.handler.builtin.controller;

import icu.takeneko.omms.central.controller.ControllerManager;
import icu.takeneko.omms.central.network.session.SessionContext;
import icu.takeneko.omms.central.network.session.handler.builtin.BuiltinRequestHandler;
import icu.takeneko.omms.central.network.session.request.Request;
import icu.takeneko.omms.central.network.session.response.Response;
import icu.takeneko.omms.central.network.session.response.Result;
import icu.takeneko.omms.central.permission.Permission;
import icu.takeneko.omms.central.util.Util;

public class ListControllersRequestHandler extends BuiltinRequestHandler {
    @Override
    public Response handle(Request request, SessionContext session) {
        var controllerNames = ControllerManager.INSTANCE.getControllers().keySet();
        var json = Util.gson.toJson(controllerNames);
        return new Response().withResponseCode(Result.CONTROLLER_LISTED).withContentPair("names", json);
    }

    @Override
    public Permission requiresPermission() {
        return null;
    }
}